### PR TITLE
mds: misc multimds fixes

### PR DESCRIPTION
**Pull request by ukernel**

### CODE REVIEW

File: src/mds/Server.cc [hunk 1]
```diff
@@ -5534,7 +5534,14 @@ void Server::handle_client_unlink(MDRequestRef& mdr)
...
     // subtree root auths need to be witnesses
     set<mds_rank_t> witnesses;
-    in->list_replicas(witnesses);
+
+    list<CDir*> dfls;
+    diri->get_subtree_dirfrags(dfls);
+    for (auto p : dfls) {
+      if (mds->get_nodeid() != p->get_dir_auth().first)
+	witnesses.insert(p->get_dir_auth().first);
+    }
```
Comment (gregsfortytwo) [inline src/mds/Server.cc]: If we don't make other replicas a witness here, at which point do they find out the inode has been unlinked?

File: src/mds/Locker.cc [hunk 2]
```diff
@@ -2049,6 +2055,10 @@ void Locker::resume_stale_caps(Session *session)
     if (cap->is_stale()) {
       dout(10) << " clearing stale flag on " << *in << dendl;
       cap->clear_stale();
+
+      if (in->state_test(CInode::STATE_EXPORTINGCAPS))
+	continue;
```
Comment (gregsfortytwo) [inline src/mds/Locker.cc]: This is subtle; can you explain in the commit message? I think this is assuming we're the recipient and can just toss out the stale caps now, right?

File: src/mds/Server.cc (hunk 1 above)
Comment (ukernel) [inline src/mds/Server.cc]: Server::_unlink_local_finish send dentry unlink message to the rest replicas

**Comment by dachary** [comment]
jenkins test this please (asok http://tracker.ceph.com/issues/15249)

**Comment by gregsfortytwo** [comment]
@ukernel, you appear to have pushed a bunch of new and unrelated bug fix commits on top of an already-reviewed PR? It'll help if you don't apply new ones once a PR has been looked at so it's clear what has and hasn't been reviewed.

File: src/mds/Migrator.h [hunk 3]
```diff
@@ -89,7 +89,6 @@ class Migrator {
     set<mds_rank_t> warning_ack_waiting;
     set<mds_rank_t> notify_ack_waiting;
     map<inodeno_t,map<client_t,Capability::Import> > peer_imported;
-    list<MDSInternalContextBase*> waiting_for_finish;
```
Comment (gregsfortytwo) [inline src/mds/Migrator.h]: "mds: remove useless code in Migrator"
With commits like that it's best to say exactly why it's useless. In this case, I gather nobody ever put anything on waiting_for_finish or called add_export_finish_waiter(), but the commit should say so.

File: src/mds/journal.cc [hunk 4]
```diff
@@ -2132,11 +2132,9 @@ void EUpdate::replay(MDSRank *mds)
 	       << " < " << cmapv << dendl;
       // open client sessions?
       map<client_t,entity_inst_t> cm;
-      map<client_t, uint64_t> seqm;
       bufferlist::iterator blp = client_map.begin();
       ::decode(cm, blp);
-      mds->server->prepare_force_open_sessions(cm, seqm);
-      mds->server->finish_force_open_sessions(cm, seqm);
+      mds->sessionmap.open_sessions(cm);
```
Comment (gregsfortytwo) [inline src/mds/journal.cc]: I don't understand why this change is being made at all. I think maybe the commit means you want to switch to doing this, but I'm not sure why. It looks like you're just tossing out a bunch of asserts and other safety/metadata code.

**Comment by ukernel** [comment]
updated

File: src/mds/journal.cc (hunk 4 above)
Comment (gregsfortytwo) [inline src/mds/journal.cc]: Okay, makes sense that trying to use the same code on both import and replay wouldn't work.

**Comment by gregsfortytwo** [comment]
Reviewed-by: Greg Farnum <gfarnum@redhat.com>
